src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,333 @@
+        }
+    }
+    return reallocation;
+}
+
+void partition_balancer_planner::get_unavailable_nodes_reassigments(
+  std::vector<ntp_reassigments>& reassigments) {
+    for (const auto& t : _topic_table.topics_map()) {
+        for (const auto& a : t.second.get_assignments()) {
+            if (_reallocation_ordered == _config.batch_size) {
+                return;
+            }
+            std::vector<model::broker_shard> stable_replicas;
+            bool need_reallocation = false;
+            for (const auto& r : a.replicas) {
+                if (!_unavailable_nodes.contains(r.node_id)) {
+                    stable_replicas.push_back(
+                      model::broker_shard{r.node_id, r.shard});
+                } else {
+                    need_reallocation = true;
+                }
+            }
+            if (need_reallocation) {
+                auto partition_size = get_partition_size(
+                  t.second.get_configuration().tp_ns.ns,
+                  t.second.get_configuration().tp_ns.tp,
+                  a.id);
+                if (
+                  partition_size == size_t(-1)
+                  || check_partition_movement_is_available(a.replicas)) {
+                    continue;
+                }
+                auto new_allocation_units = get_reallocation(
+                  a, t.second, stable_replicas, partition_size);
+                if (
+                  !new_allocation_units
+                  || new_allocation_units.value().get_assignments().size()
+                       == 0) {
+                    vlog(
+                      clusterlog.info,
+                      "Can't reallocate {} {} {} from down node",
+                      t.first.ns,
+                      t.first.tp,
+                      a.id);
+                } else {
+                    reassigments.emplace_back(ntp_reassigments{
+                      .ntp = model::ntp(t.first.ns, t.first.tp, a.id),
+                      .allocation_units = std::move(
+                        new_allocation_units.value())});
+                    _reallocation_ordered += partition_size;
+                }
+            }
+        }
+    }
+}
+
+struct node_space {
+    model::node_id node_id;
+    float free_space_percent;
+    uint64_t free_space;
+    uint64_t total_space;
+};
+
+bool node_space_sorter(node_space const& lhs, node_space const& rhs) {
+    return lhs.free_space_percent < rhs.free_space_percent;
+}
+
+void partition_balancer_planner::get_full_node_reassigments(
+  std::vector<ntp_reassigments>& reassigments) {
+    std::vector<node_space> nodes_with_disk_constaints_violation;
+    for (const auto& n : _node_reports) {
+        for (const auto& disk : n.second.local_state.disks) {
+            float available_space = float(disk.free) / float(disk.total);
+            if (available_space < float(1) - _config.max_disc_usage) {
+                nodes_with_disk_constaints_violation.emplace_back(node_space{
```

Comment:
why not make it a constructor so that struct initialization is invalid and the user must use valid object construction.